tutorial/testing.py

```diff
@@ -0,0 +1,295 @@
+    [pytest](https://docs.pytest.org/en/latest/) and
+    [selenium webdriver](https://www.seleniumhq.org/projects/webdriver/),
+    but focuses on how to do a simple integration test with dash by hosting
+    the app server locally and using a selenium webdriver to simulate
+    the interaction inside a web browser.
+
+    ![demo](https://user-images.githubusercontent.com/1394467/59383731-83c13480-8d2e-11e9-8866-4ffdcd3b1b45.gif)
+
+    ## Install
+
+    The dash testing is now part of the main dash package.  After
+    `pip install dash`, the *dash pytest* fixtures are available, you just
+    need to install the webdrivers and you are ready to test.
+
+    - [Chrome Driver](http://chromedriver.chromium.org/getting-started)
+    - [Firefox Gecko Driver](https://github.com/mozilla/geckodriver/releases)
+
+    FYI, We run dash integration tests with `Chrome` webdriver.
+    But the fixture allows you to choose another browser from the command line,
+    e.g. `pytest --webdriver Firefox -k bsly001`.
+
+    **Notes**:
+
+    * The *Gecko(Marionette)* driver from Mozilla is not fully compatible with
+    selenium specifications. Some features may not work as expected.
+
+    * We only include *Chrome* and *Firefox* in the supported list for now,
+    but other popular webdrivers can be included if there are popular demands.
+
+    ## Example
+    """)),
+    Syntax(code),
+    dcc.Markdown(dedent("""
+
+    **Notes**:
+
+    * #1 For most test scenarios, you don't need to import any modules for
+    the test, just import what you need for the dash app itself.
+
+    * #2 A test case is a regular python function, the function name follows
+    this pattern `test_{tcid}_{test title}`. The `tcid` (test case ID) is
+    an abbreviation that follows `mmffddd => module + file + three digits`.
+    The `tcid` facilitates the test selection by just running
+    `pytest -k {tcid}`. Its naming convention also helps code navigation with
+    modern editors.
+
+    * #3 Here we just define our app inside a test function.
+    All the rules apply as in your app file.
+
+    * #4 We normally start the test by calling the `start_server` API
+    from dash_duo. Several actions implicitly happen under the hood:
+
+        1. It hosts the defined app with Flask in a light python
+        `threading.Thread`
+        2. A selenium webdriver is initialized and it navigates to the
+        local server with the `server_url`
+        3. It waits until the page has loaded in the browser
+
+    * #5 A test case is composed of preparation, actions and check points.
+    Both #5 and #6 are doing the same check in this example, we are expecting
+    that the defined `Div` component's text is identical to the `children`'s
+    content. #5 will wait the expected state to reach within an explicitly 4
+    seconds timeout. It's a safer way to use when you are doing an element
+    check related to callbacks, as it happens at lot under dash context,
```

Comment:
```suggestion
    check related to callbacks, as it is usually performed in the Dash context;
```

(not sure if "usually" is what you meant by "a lot")